### ISSUE TITLE
Long compile times due to initialization of large struct 

### ISSUE THREAD

**Issue by fintelia** [rust-lang/rust#52960]
<short summary of the bug>

I tried fixing erlepereira/x11-rs#90 with [this patch](https://github.com/fintelia/x11-rs/commit/7d6ee99a5482f3cd5b2eaa1a1402e36ec22111a0) that directly instantiates some (very) large structs via a macro. Unfortunately this change caused compile times to drastically increase. Before a build took only a few seconds, but now a debug build takes 40+ seconds and my attempt to do a release build has been going for ~30 minutes~ over an hour.

[~/git/x11-rs/x11-dl$ cargo rustc -- -Z time-passes](https://pastebin.com/DXfRdxkD)

[~/git/x11-rs/x11-dl$ cargo rustc --release -- -Z time-passes](https://pastebin.com/NyVfxSVQ) (incomplete log output, killed after producing no further output for an hour.)

## Meta

rustc --version --verbose:
```
rustc 1.29.0-nightly (e94df4acb 2018-07-31)
binary: rustc
commit-hash: e94df4acb4c3f42fdc224a7164b63a99240add1e
commit-date: 2018-07-31
host: x86_64-unknown-linux-gnu
release: 1.29.0-nightly
LLVM version: 7.0
```

**Comment by fintelia** [comment]
Yes, the macros do generate a lot of code, but it isn't *that* much code. Cargo expand produces under 50k lines of code, and none of it looks all that complex.

The issue seems to be related to having a large number of control flow paths through the function. If I change the `try!` to an `unwrap()` compile time falls to under 10 seconds.

**Comment by carado** [comment]
The compilation might not be *just* long, it might be running in an infinite loop or at least something exponential.
I tried compiling my project with the original x11-dl replaced with the patched version mentioned above and letting it compile overnight (~8 hours) and it didn't finish compiling when I woke up.
The attempt was a compiling in debug mode, with the following options:
```
[profile.dev]
debug = false
opt-level = 2
```

**Comment by ishitatsuyuki** [comment]
Marking as LLVM since opt-level matters here and that is often caused by conflicting LLVM optimizations.

**Comment by fintelia** [comment]
Just checked the original commit again and it no longer repros. Seems to have been fixed 5+ years ago, so I didn't bother bisecting further
